File: commands/games.ts
import axios, { AxiosResponse } from "axios";
import User from "../models/User";
import randomcolor from "randomcolor";
import shuffle from "shuffle-array";

import { RichEmbed, GuildMember, TextChannel, Message } from "discord.js";
import { User as User_ } from "discord.js";

interface TriviaUser extends User_ {
  trivia?: boolean;
  answered?: boolean;
}

export default {
  dice: {
    args: "dice <numberOfDice>d<sides>",
    description: "roll <numberOfDice> of dice that have <sides> sides",
    async use(
      member: GuildMember,
      channel: TextChannel,
      message: Message,
      args: string[],
    ): Promise<Message | Message[]> {
      const numberOfDice = Number(args[0].split("d")[0]);
      const sides = Number(args[0].split("d")[1]);
      const results: number[] = [];
      for (let i = 0; i < numberOfDice; i++) {
        const result: number =
          Math.floor(Math.random() * Math.floor(sides)) + 1;
        results.push(result);
      }
      return channel.send(
        `:game_die: Roll ${numberOfDice} ${sides}-sided dice. :game_die: \n:game_die: Results: ${results.join(
          ", ",
        )}. Total: ${results.reduce((x, y) => x + y)}.:game_die:`,
      );
    },
  },

  points: {
    args: "points | points <user>",
    description: "get points of yourself, or <user>",
    async use(
      member: GuildMember,
      channel: TextChannel,
      message: Message,
      args: string[],
    ): Promise<Message | Message[]> {
      if (message.mentions.members.size > 0) {
        const discordUser: GuildMember = message.mentions.members.first();
        const user: User = await User.findOrCreate(discordUser.id);
        return channel.send(
          `${discordUser.user.username} has ${user.points} points`,
        );
      } else {
        const user: User = await User.findOrCreate(member.user.id);
        return channel.send(
          `${member.user.username} has ${user.points} points`,
        );
      }
    },
  },
  trivia: {
    args: "trivia",
    description: "posts a trivia question using **opentdb** API",
    async use(
      member: GuildMember,
      channel: TextChannel,
      message: Message,
      args: string[],
    ): Promise<void> {
      const user = await User.findOrCreate(member.user.id);

      axios
        .get(
          "https://opentdb.com/api.php?amount=1&type=multiple&encode=url3986",
        )
        .then((res: AxiosResponse) => {
          const data = res.data.results[0];
          const question: string = data.question.replace(/&quot;/g, '"');
          const triviaUser: TriviaUser = member.user;
          triviaUser.trivia = true;
          const choices: string[] = Object.values(data.incorrect_answers);
          choices.push(data.correct_answer);
          shuffle(choices);
          choices.map((i) => i.replace(/&#039;/g, "'").replace(/&quot;/g, '"'));
          const difficulty: string = data.difficulty;
          const correctAnswerNumber: number =
            choices.indexOf(data.correct_answer) + 1;
          const correctAnswer: string = data.correct_answer;
          const embed: RichEmbed = new RichEmbed()
            .setTitle(`${triviaUser.username}'s Question`)
            .setAuthor(triviaUser.username, triviaUser.avatarURL)
            .addField("Difficulty", difficulty)
            .addField("Question (You Have 20 Seconds To Answer)", question)
            .addField("1)", choices[0], true)
            .addField("2)", choices[1], true)
            .addBlankField()
            .addField("3)", choices[2], true)
            .addField("4)", choices[3], true)
            .setColor(randomcolor());

          channel.send(embed);
          channel
            .awaitMessages((m) => m.author === member.user, {
              max: 1,
              time: 20000,
              errors: ["time"],
            })
            .then(async (m) => {
              triviaUser.answered = true;
              if (Number(m.array()[0].content) != correctAnswerNumber) {
                channel.send(`Incorrect, the answer is ${correctAnswer}.`);
              } else {
                channel.send("Correct!");
                user.points +=
                  difficulty === "easy" ? 5 : difficulty === "medium" ? 10 : 15;
                await user.save();
              }
            })
            .catch(() => {
              if (!triviaUser.answered)
                channel.send(`Time's up! The answer was ${correctAnswer}.`);
            });
          triviaUser.answered = false;
        });
    },
  },
};

File: commands/games.py

import random
from db import Session
from models.user import User
session = Session()
from discord import Embed
import requests
import asyncio

async def dice(client, member, channel, message, args):
    number_of_dice = int(args[0].split("d")[0])
    number_of_sides = int(args[0].split("d")[1])

    results = [random.randint(1, number_of_sides) for i in range(1, number_of_dice+1)]
    await channel.send(results)

async def points(client, member, channel, message, args):
    user = session.query(User).filter_by(discord_id=member.id).first()
    points = None
    if not user:
        user = User(discord_id=member.id, points=0)
        session.add(user)
        session.commit()
        points = 0
    else:
        points = user.points

    await channel.send(f"You have {points} {'Point' if points == 1 else 'Points'}.")


async def trivia(client, member, channel, message, args):
    url = "https://opentdb.com/api.php?amount=1&type=multiple&encode=url3986"
    data = requests.get(url).json()["results"][0]

    question = data["question"]
    #member["trivia"] = True
    choices = data["incorrect_answers"]
    print(choices)
    choices.append(data["correct_answer"])
    difficulty = data["difficulty"]
    correct_answer = choices.index(data["correct_answer"]) + 1

    embed = Embed(title=f"{member.name}'s Question")

    embed.set_author(name=member.name,icon_url=member.avatar_url)
    embed.add_field(name="Difficulty",value=difficulty, inline=False)
    embed.add_field(name="Question (You Have 20 Seconds To Answer)", value=question, inline=False)
    embed.add_field(name="1)", value=choices[0], inline=True)
    embed.add_field(name="2)", value=choices[1], inline=True)
    embed.add_field(name="3)", value=choices[2], inline=False)
    embed.add_field(name="4)", value=choices[3], inline=True)

    await channel.send(embed=embed)

    def check(m):
        return m.author == member

    try:
        response = await client.wait_for("message", check=check, timeout=5.0)
        print(response)
    except asyncio.TimeoutError:
        await channel.send("no response.")

